### PR TITLE
Correct faulty setting of Capabilities for writable SpatialViews.

### PR DESCRIPTION
**Pull request by mj10777**
## Description
Correct faulty setting of Capabilities for writable SpatialViews.
When searching for triggers of a writable SpatialView, no checking was being done to check the type of triggers found.

The new checking function will check for the INSERT, UPDATE and DELETE triggers and set the Capabilities as needed.

A test has been added to check an existing Database with 3 SpatialViews
* one with all 3 Capabilities
* one with only INSERT and UPDATE Capabilities
* one with only an INSERT Capabilitiy

which checks that each SpatialViews has the proper Capabilities.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by mj10777** [comment]
As far as I can see, the Travis errors have nothing to do with this pull request.

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 1]
```diff
@@ -529,7 +532,7 @@ QgsSpatiaLiteProvider::QgsSpatiaLiteProvider( QString const &uri )
     return;
   }
   mEnabledCapabilities = mPrimaryKey.isEmpty() ? QgsVectorDataProvider::Capabilities() : ( QgsVectorDataProvider::SelectAtId );
-  if ( ( mTableBased || mViewBased ) &&  !mReadOnly )
+  if ( mTableBased )
```
Comment (m-kuhn) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Should that be `if ( mTableBased && !mReadOnly )`? Readonly table based connections have previously not been handled in this block and now are

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 2]
```diff
@@ -540,6 +543,26 @@ QgsSpatiaLiteProvider::QgsSpatiaLiteProvider( QString const &uri )
     mEnabledCapabilities |= QgsVectorDataProvider::AddAttributes;
     mEnabledCapabilities |= QgsVectorDataProvider::CreateAttributeIndex;
   }
+  if ( mViewBased  &&  !mReadOnly )
```
Comment (m-kuhn) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Could be an `else if` I guess

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 1 above)
Comment (mj10777) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: The ReadOnly applies only to views in Spatialite.
Here, I believe, it is only used together with views or VirtualShapes (which are always readonly).
There are no triggers for TABLES that make them writable.

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 2 above)
Comment (mj10777) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Yes, I have added that.

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 3]
```diff
@@ -5712,6 +5768,9 @@ QGISEXTERN QString loadStyle( const QString &uri, QString &errCause )
   QgsDataSourceUri dsUri( uri );
   QString sqlitePath = dsUri.database();
   QgsDebugMsg( "Database is: " + sqlitePath );
+  // Avoid sql-statement errors
+  if ( dsUri.schema().isEmpty() )
+    return false;
```
Comment (m-kuhn) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: the method returns a `QString`

Comment (mj10777) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Where is this being used?
When dsUri.schema() is empty, it caused sql-errors.
The same for saveStyle.

Comment (m-kuhn) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Have a look at how the `if ( !handle )` on line 5777 deals with error reporting (and better use `QString()` than the empty `QLatin1String( "" )` there)

Comment (mj10777) [inline src/providers/spatialite/qgsspatialiteprovider.cpp]: Sorry, made the correction in the wrong copy.
The 'else' (from above) has been added and QString()  returned for both cases with a message for the reason.

**Comment by mj10777** [comment]
Starting with Spatialite 4.5.0, gaiaGetVectorLayersList will return information about the Capabilities of writable Views in the same way as added here.

**Pull request closed without merging**